x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/action/TransportFreezeIndexAction.java

```diff
@@ -0,0 +1,221 @@
+                                      ThreadPool threadPool, ActionFilters actionFilters,
+                                      IndexNameExpressionResolver indexNameExpressionResolver,
+                                      DestructiveOperations destructiveOperations) {
+        super(FreezeIndexAction.NAME, transportService, clusterService, threadPool, actionFilters, indexNameExpressionResolver,
+            FreezeRequest::new);
+        this.destructiveOperations = destructiveOperations;
+    }
+    @Override
+    protected String executor() {
+        return ThreadPool.Names.SAME;
+    }
+
+    @Override
+    protected void doExecute(Task task, FreezeRequest request, ActionListener<AcknowledgedResponse> listener) {
+        destructiveOperations.failDestructive(request.indices());
+        super.doExecute(task, request, listener);
+    }
+
+    @Override
+    protected AcknowledgedResponse newResponse() {
+        return new AcknowledgedResponse();
+    }
+
+    @Override
+    protected void masterOperation(FreezeRequest request, ClusterState state, ActionListener<AcknowledgedResponse> listener) {
+        final Index[] concreteIndices = indexNameExpressionResolver.concreteIndices(state, request);
+        if (concreteIndices == null || concreteIndices.length == 0) {
+            throw new ResourceNotFoundException("index not found");
+        }
+
+        clusterService.submitStateUpdateTask("toggle-frozen-settings",
+            new AckedClusterStateUpdateTask<AcknowledgedResponse>(Priority.URGENT, request, listener) {
+            @Override
+            public ClusterState execute(final ClusterState currentState) {
+                final MetaData.Builder builder = MetaData.builder(currentState.metaData());
+                ClusterBlocks.Builder blocks = ClusterBlocks.builder().blocks(currentState.blocks());
+                for (Index index : concreteIndices) {
+                    IndexMetaData meta = currentState.metaData().getIndexSafe(index);
+                    if (meta.getState() != IndexMetaData.State.CLOSE) {
+                        throw new IllegalStateException("index [" + index.getName() + "] is not closed");
+                    }
+                    final IndexMetaData.Builder imdBuilder = IndexMetaData.builder(meta);
+                    final Settings.Builder settingsBuilder =
+                        Settings.builder()
+                            .put(currentState.metaData().index(index).getSettings())
+                            .put("index.blocks.write", request.freeze())
+                            .put(FrozenEngine.INDEX_FROZEN.getKey(), request.freeze())
+                            .put(IndexSettings.INDEX_SEARCH_THROTTLED.getKey(), request.freeze());
+                    if (request.freeze()) {
+                        blocks.addIndexBlock(index.getName(), IndexMetaData.INDEX_WRITE_BLOCK);
+                        // we never remove this block when unfreeze for now. we don't know if it was read-only
```

Comment:
I think this should be removed based on the value of `index.blocks.write` (i.e., if true add, if false remove). See `MetaDataUpdateSettingsService#updateSettings`.